stored_file.php

```diff
@@ -0,0 +1,387 @@
+            return static::resolve_file($this->filename);
+        return $this->source;
+    }
+
+    public function has_supported_extension($supported_exts = null)
+    {
+        if (!$this->extension)
+            $this->extension = static::parse_extension($this->filename);
+
+        if (!$this->extension)
+            return false;
+
+        if ($supported_exts && is_array($supported_exts) && count($supported_exts))
+        {
+            return in_array($this->extension,
+                $supported_exts);
+        }
+
+        return in_array($this->extension,
+            static::$__supported_exts);
+    }
+
+    public function exists()
+    {
+        return is_file($this->actual_filename());
+    }
+
+    public function size()
+    {
+        return filesize($this->actual_filename());
+    }
+
+    public function human_size()
+    {
+        $size = $this->size();
+
+        if ($size > static::SIZE_GIB)
+        {
+            $size = ($size / static::SIZE_GIB);
+            return sprintf('%.2f GiB', $size);
+        }
+
+        if ($size > static::SIZE_MIB)
+        {
+            $size = ($size / static::SIZE_MIB);
+            return sprintf('%.2f MiB', $size);
+        }
+
+        $size = ($size / static::SIZE_KIB);
+        return sprintf('%.2f KiB', $size);
+    }
+
+    public function detect_mime()
+    {
+        if ($this->mime) return $this->mime;
+        $this->mime = File_Util::detect_mime($this->actual_filename());
+        return $this->mime;
+    }
+
+    public function generate_filename($ext = null)
+    {
+        if (!$ext) $ext = $this->extension;
+
+        $hash = md5_file($this->source);
+        $dir1 = substr($hash, 0, 2);
+        $dir2 = substr($hash, 2, 2);
+        $name = substr($hash, 4, 28);
+
+        $this->filename = build_path($dir1, $dir2,
+            sf('%s.%s', $name, $ext));
+        return $this->filename;
+    }
+
+    public function read()
+    {
+        return file_get_contents($this->actual_filename());
+    }
+
+    public function move($invoker_tag = null)
+    {
+        if ($this->__moved) return;
+
+        if (!$invoker_tag) $invoker_tag = md5(serialize(debug_backtrace(5)));
+
+        $destination = static::resolve_file($this->filename, true);
+        $dir = dirname($destination);
+        $ref_filename = $destination . '.ref';
+
+        if (!is_dir($dir)) mkdir($dir, 0755, true);
+
+        if (!file_exists($destination)) {
+
+            rename($this->source, $destination);
+
+            // Create a new file reference
+            $ref_res = fopen($ref_filename, 'w');
+            if (flock($ref_res, LOCK_EX)) {
+                ftruncate($ref_res, 0);
+                fwrite($ref_res, "{\"ref_count\": 1,\"atags\" : [\"$invoker_tag\"],\"dtags\" : []}");
+                fflush($ref_res);
+                flock($ref_res, LOCK_UN);
+                fclose($ref_res);
+            } else {
+                throw new \RuntimeException("Couldn't get the lock!");
```

Comment:
Why was there no attempt to try again?

Reply:
Because none of the methods that use flock take more than 10ms to release a locked file, so if we have an issue related to locked files means that one of the PHP scripts was abruptly interrupted. If you really need a Strong Consistency then I can isolate the locking functionality into a class that can control the retries (using a while statement) and register a shutdown function to release currently locked files, or just use a DB table as I proposed initially.